include/alpaka/rand/RandDefault.hpp

```diff
@@ -0,0 +1,198 @@
+ */
+
+#pragma once
+
+#include <alpaka/core/Common.hpp>
+#include <alpaka/core/Unused.hpp>
+#include <alpaka/math/cos/Traits.hpp>
+#include <alpaka/math/log/Traits.hpp>
+#include <alpaka/math/sin/Traits.hpp>
+#include <alpaka/math/sqrt/Traits.hpp>
+#include <alpaka/rand/RandPhilox.hpp>
+#include <alpaka/rand/Traits.hpp>
+
+#include <limits>
+#include <type_traits>
+
+namespace alpaka
+{
+    namespace rand
+    {
+        class RandDefault : public concepts::Implements<ConceptRand, RandDefault>
+        {
+        };
+
+        namespace distribution
+        {
+            namespace gpu
+            {
+                namespace detail
+                {
+                    template<typename TFloat>
+                    struct BitsType;
+
+                    template<>
+                    struct BitsType<float>
+                    {
+                        typedef std::uint32_t type;
+                    };
+                    template<>
+                    struct BitsType<double>
+                    {
+                        typedef std::uint64_t type;
+                    };
+                } // namespace detail
+
+                //! The GPU random number normal distribution.
+                template<typename T>
+                class UniformUint
+                {
+                public:
+                    UniformUint() = default;
+
+                    template<typename TEngine>
+                    ALPAKA_FN_HOST_ACC auto operator()(TEngine& engine) -> T
+                    {
+                        using BitsT = typename TEngine::result_type;
+                        T ret = 0;
+                        constexpr auto N = sizeof(T) / sizeof(BitsT);
+                        for(unsigned int a = 0; a < N; ++a)
+                        {
+                            ret ^= (static_cast<T>(engine()) << (sizeof(BitsT) * 8 * a));
+                        }
+                        return ret;
+                    }
+                };
+
+                //! The GPU random number uniform distribution.
+                template<typename T>
+                class UniformReal
+                {
+                    using BitsT = typename detail::BitsType<T>::type;
+
+                public:
+                    UniformReal() = default;
+
+                    template<typename TEngine>
+                    ALPAKA_FN_HOST_ACC auto operator()(TEngine& engine) -> T
+                    {
+                        constexpr BitsT limit = static_cast<BitsT>(1) << std::numeric_limits<T>::digits;
+                        const BitsT b = UniformUint<BitsT>()(engine);
+                        const auto ret = static_cast<T>(b & (limit - 1)) / limit;
+                        return ret;
+                    }
+                };
+
+                //! The GPU random number normal distribution.
+                template<typename Acc, typename T>
+                class NormalReal
+                {
+                    const Acc& m_acc;
```

Comment:
IMO the copy operator of `NormalReal` should be deleted to avoid that the user is storing this object in global data. Maybe adding a comment to oxygen of this class could also help.

Reply:
Technically, copy and move are already implicitly deleted because `NormalReal` contains a reference.

Reply:
Should we explicitly delete this copy and/or move constructor?

Reply:
I am not sure how this really keeps the user form storing it in global memory and sharing it (good point about the comment).

Deleting copy would avoid someone copying the cache and possibly getting two times the same value. Not other distribution has deleted copies, so this would make this one different. If the copy is not deleted, it should not be trivial and clear the cache of the copy, but this would add overhead which is not there for other distributions. As it stands deleting copy could break generic code, so I would avoid this unless all distributions' copies are deleted (same for move.)

Reply:
Ok, let's be precise: the copy and move assignment operators are deleted. The constructors should not be impacted.

> The compiler just turns the reference into a pointer.

This happens in the codegen phase. But syntactically, you still have a reference which restricts the surrounding class.

Reply:
Deleting both move and copy breaks the tests unless we move to C++17 where copy elision is mandatory for return values.

Reply:
> Should we explicitly delete this copy and/or move constructor?

Even if you delete them, you could still placement new a `NormalReal` into a buffer. And IMO there is nothing wrong with copying/moving that class around. So I would even change the reference member to a pointer member, so the class can be copied/moved. But we should document that this class keeps a reference to the passed accelerator and that an instance should not outlive a kernel invocation.

Reply:
@bernhardmgruber was referring to the copy/move operators there, those are implicitly deleted.

Reply:
Pushed https://github.com/alpaka-group/alpaka/pull/1444/commits/a8b261fe42734112e484216522c5a436fdb27191 which cleanly implements copy and move ctors.

Reply:
> Deleting both move and copy breaks the tests unless we move to C++17 where copy elision is mandatory for return values.

Please also turn the stored accelerator object into a pointer, so the class can be freely copied. That should make the tests work.

Reply:
> And IMO there is nothing wrong with copying/moving that class around.
> So I would even change the reference member to a pointer member, so the class can be copied/moved.

There is no harm as such, but for performance it would be best to pass a reference to it around. Both copying and creating multiple instance only wastes registers on GPU. Letting an instance go out of scope and creating a new one when needed again also saves registers and is cheaper than a copy. I do not think copy and assignment of this type, or any alpaka distribution is useful, as they do not contain parameters (like min/max/sigma/mean).


> But we should document that this class keeps a reference to the passed accelerator and that an instance should not outlive a kernel invocation.

Ok. This is formally true for any object created from an `Acc`, i.e. all distributions and also random engines, this is just the only one that uses its `Acc` for something.

Reply:
Github issues do not appear to work when we write simultaneously. Your comments only appear to me somewhere in the middle long after I wrote something else, so I miss them. Then I get an email about a comment which turns out to be in the middle of the conversation... .

> > The compiler just turns the reference into a pointer.
> 
> This happens in the codegen phase. But syntactically, you still have a reference which restricts the surrounding class.

Yes, this is the intention: You cannot assign to a reference. `m_acc` should not change during the lifetime of the object, anything else would be a bug, hence the reference.

We could actually add assignments that do not touch the reference. See https://github.com/alpaka-group/alpaka/pull/1444/commits/31d18e754579a9ec05dd6ad71df443d293a2abb3 . This is the semantically clean way to do the assignments, though it is admittedly pointless. Note, that these object always exist withing a kernel, where there is only one `Acc` handle, i.e. you can never assign an object with a different `m_acc`.

Reply:
> Both copying and creating multiple instance only wastes registers on GPU. Letting an instance go out of scope and creating a new one when needed again also saves registers and is cheaper than a copy.

I strongly trust the optimizer to inline everything and remove all redundancies. Passing references around also costs registers and AFAIK CUDA cannot even form pointers/references into registers, so the optimizer must eliminate those anyway. So in the end it should not matter whether you copy distribution objects or pass them by references. And since the object is light-weight, I would prefer making it as regular as possible. But after this discussion, I don't care so much anymore. You are free to choose a behavior.

Reply:
The CUDA compiler will inline everything and remove reference function arguments in the process, as they are just the same variable. If you copy, in the case for this `NormalReal` is not redundant because it contains the cache variable which changes independently for original and copy in their respective `operator()` calls, thus the copy cannot be optimized out (I agree, that this will be different for all other distribution, because they do not have any data members). This behavior does not depend on how exactly the copy ctor or assignment behave.

Going with deleting as much as possible, to try keep the user from getting unexpected results at runtime from this class.